config-service-change-event-api/src/main/java/org/hypertrace/config/change/event/v1/ConfigChangeEventKey.java

```diff
@@ -0,0 +1,41 @@
+package org.hypertrace.config.change.event.v1;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import java.util.Objects;
+
+/** Key for config change event */
+public class ConfigChangeEventKey {
```

Comment:
Yes, the consumer can determine from the key which config is being updated. They also need this information to deserialize.

Reply:
Since you pointed out issue with using protobuf for key, I went with json for keys.

Reply:
One point I overlooked in the page you had shared -
```
The indeterminism is caused by -
1. the absence of guarantees on how known or unknown fields should be ordered upon writing them
2. map serialisation relies on iteration, and there is no ordering guarantee while iterating over map items
```
A simple proto with string values will never run in to above issue. Based on this I am pretty confident in exposing key as proto itself.

Reply:
> A simple proto with string values will never run in to above issue. Based on this I am pretty confident in exposing key as proto itself.

"how known fields should be ordered" would be the issue that could apply here, since we're (at the moment) sending all fields and not using maps. That still leaves risk in evolution though, if a new field were to be added.

But assuming it works, the bigger risk we should consider IMO - will the next person realize these considerations when they reference this to set up a new topic with proto keys? We just want to be careful when establishing a pattern.

So, a summary of my thoughts:
- The point about bringing up proto concerns for keys was to consider and mitigate any issues, but doesn't mean we shouldn't use them necessarily
- Using proto gives us multi-language support, backwards compatibility/breaking change detection support and is our standard elsewhere, so should be our default unless we have serious reasons not to use it
- JSON as a transport format doesn't give us much advantage, as we could just as easily represent a proto key in json. It's only a question of getting a consistent serializer, or determining that the off the shelf proto serializer is good enough for our needs 
- For the vast majority of our usages, we have one publisher and N subscribers (more often than not one, but sometimes more). At least, I can't think of any topics today where we use multiple publishers to one topic, although it's a perfectly valid topology. The concern about key consistency is on the serialization side, which seems at least partially mitigated by using a single publisher unless the serializer is self-inconsistent.
- The same thought process around having a more complex key seems like it goes for the entity change topic too - the consumer should be able to read the type off the key before requiring deserializing the values. That is, give me only API changes for example - and we don't want magic concatenated strings if we need to read them.